enhancements/network/baremetal-ipi-network-configuration.md

```diff
@@ -0,0 +1,187 @@
+  - "@kirankt"
+  - "@dtantsur"
+  - "@zaneb"
+approvers:
+  - "TBD"
+creation-date: 2021-05-21
+last-updated: 2021-06-28
+status: provisional|implementable
+
+see-also:
+  - "/enhancements/host-network-configuration.md"
+  - "/enhancements/machine-config/mco-network-configuration.md"
+  - "/enhancements/machine-config/rhcos/static-networking-enhancements.md"
+
+# Baremetal IPI Network Configuration
+
+Describe user-facing API for day-1 network customizations in the IPI workflow,
+with particular focus on baremetal where such configuration is a common
+requirement.
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Operational readiness criteria is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+Currently in the IPI flow, there is no way to provide day-1 network configuration
+which is a common requirement, particularly for baremetal users.  We can build
+on the [UPI static networking enhancements](rhcos/static-networking-enhancements.md)
+to enable such configuration in the IPI flow.
+
+## Motivation
+
+Since the introduction of baremetal IPI, a very common user request is how
+to configure day-1 networking, and in particular the following cases which are not currently possible:
+
+* Deploy with OpenShift Machine network on a tagged (non-default) VLAN
+* Deploy with OpenShift Machine network using static IPs (no DHCP)
+
+In both cases, this configuration cannot be achieved via DHCP so some
+means of providing the configuration to the OS is required.
+
+In the UPI flow this is achieved by consuming user-provided NetworkManager
+keyfiles, as an input to `coreos-install --copy-network`, but there is
+no corresponding user interface at the openshift-install level.
+
+### Goals
+
+* Define API for day-1 network customizations
+* Enable common on-premise network configurations (bond+vlan, static ips) via IPI
+
+### Non-Goals
+
+* Platforms other than `baremetal`, although the aim is a solution which could be applied to other platforms in future if needed.
+* Provide a consistent (ideally common) user API for deployment and post-deployment configuration. Getting agreement on a common API for day-1 and day-2 has proven contentious and will be addressed later.
+* Enabling kubernetes-nmstate by default for day-2 networking is discussed via
+[another proposal](https://github.com/openshift/enhancements/pull/747)
+* Handling of advanced networking options such as bridges and bonds. The implementation may support them, but they are not the priority for this work.
+
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As a baremetal IPI user, I want to deploy via PXE and achieve a highly
+available Machine Network configuration in the most cost/space effective
+way possible.
+
```

Comment:
Allowing users to statically assign addresses in the provisioning DHCP server is probably possible, but I'm not sure it's worth tackling at this time. Users who are uncomfortable with us assigning the provisioning addresses can use virtualmedia and skip DHCP entirely. The implementation of that would also be completely separate from what we're currently proposing. It wouldn't be configuration applied to the nodes, it would be configuration applied to the provisioning DHCP server.

So for the initial implementation I would be inclined to only focus on your first and third cases. If there's significant requests for the second option we can always implement that in the future.

Reply:
IMO we should treat the provisioning network as out-of-scope for this PR - there are already ways for customers to control that network (either by disabling it and using virtual-media, or by making that network [Unmanaged](https://docs.openshift.com/container-platform/4.7/installing/installing_bare_metal_ipi/ipi-install-installation-workflow.html#additional-install-config-parameters_ipi-install-configuration-files) and configuring their own DHCP server)

Reply:
@cybertron perhaps add configuration of the provisioning network as an explicit non-goal?

Reply:
Ack, that makes sense.